### PR TITLE
New BitBucket implementation

### LINKED ISSUE #234: Netlify CMS on Bitbucket
While trying to use Netlify CMS on a website that uses Bitbucket to host it's coding, I received the following error while deploying. Chris suggested me to open an issue in this repo that will even help to develop this CMS. So, here's the first issue I faced while deploying the CMS:

CMD - Hugo Server
Started building sites ...
ERROR: 2017/02/10 14:52:37 shortcode.go:288: Unable to locate template for shortcode 's' in page extending
ERROR: 2017/02/10 14:52:37 shortcode.go:288: Unable to locate template for shortcode 's' in page index

Which is basically same as on Netlify's Deployment logs but with more specified error.
2:36:24 PM: ERROR: 2017/02/10 09:06:24 shortcode.go:288: Unable to locate template for shortcode 's' in page extending
2:36:24 PM: ERROR: 2017/02/10 09:06:24 shortcode.go:288: Unable to locate template for shortcode 's' in page index
2:36:24 PM: ERROR: 2017/02/10 09:06:24 general.go:212: Error while rendering page admin/example/index.html: html/template:__admin/example/index.html: ends in a non-text context: {stateJSRegexp delimNone urlPartNone jsCtxRegexp attrNone elementScript <nil>}

What can be done for this error?
Just as prior information, I installed Netlify CMS yesterday but the timestamp on local files is 2017/02/10. Moreover, deleting the /admin folder does removes this error.

### PR DESCRIPTION
**Pull request by Benaiah**
Closes #234.

**- Summary**

Adds support for BitBucket as a backend, both with and without git-gateway. (git-gateway support depends on the merge and deployment of https://github.com/netlify/git-gateway/pull/20). Includes token refresh logic, so users won't have to log in every couple hours to keep using the CMS.

**- Test plan**

Tested manually.

**- Description for the changelog**

Add BitBucket backend

**- A picture of a cute animal (not mandatory but encouraged)**

image

### CODE REVIEW

**Comment by verythorough** [comment]
Deploy preview for *netlify-cms-www* ready!

Built with commit 9250265cd8ad716130f9dd13765cbd8ae3dc10ce

https://deploy-preview-1504--netlify-cms-www.netlify.com

**Comment by verythorough** [comment]
Deploy preview for *cms-demo* ready!

Built with commit 9250265cd8ad716130f9dd13765cbd8ae3dc10ce

https://deploy-preview-1504--cms-demo.netlify.com

File: src/backends/git-gateway/implementation.js [hunk 1]
```diff
@@ -55,19 +58,27 @@ export default class GitGateway {
...
   }
+
+  requestFunction = req => this.tokenPromise()
+    .then(token => unsentRequest.withHeaders({ Authorization: `Bearer ${ token }` }, req))
+    .then(unsentRequest.performRequest);
+
   authenticate(user) {
-    this.tokenPromise = user.jwt.bind(user);
+    // this.tokenPromise = user.jwt.bind(user);
+    this.tokenPromise = () => Promise.resolve("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyX21ldGFkYXRhIjp7fSwiYXBwX21ldGFkYXRhIjp7InJvbGVzIjpbImFkbWluIl19fQ.5W7AGwZ8h5Y1YfHXhaMO1ysviOrKJhcCHGd4SSZ316E");
```
Comment (erquhart) [inline src/backends/git-gateway/implementation.js]: Need to switch this one back.

File: src/backends/git-gateway/implementation.js [hunk 2]
```diff
@@ -55,19 +58,27 @@ export default class GitGateway {
...
         }).then(res => res.json());
         this.acceptRoles = roles;
         if (github_enabled) {
           this.backendType = "github";
         } else if (gitlab_enabled) {
           this.backendType = "gitlab";
-        }
+        } else if (bitbucket_enabled) {
+	  this.backendType = "bitbucket";
+	}
```
Comment (erquhart) [inline src/backends/git-gateway/implementation.js]: Nitpick: fix indentation

File: src/backends/git-gateway/implementation.js [hunk 3]
```diff
@@ -98,6 +109,14 @@ export default class GitGateway {
       } else if (this.backendType === "gitlab") {
         this.api = new GitLabAPI(apiConfig);
         this.backend = new GitLabBackend(this.config, { proxied: true, API: this.api });
+      } else if (this.backendType === "bitbucket") {
+	this.api = new BitBucketAPI({
```
Comment (erquhart) [inline src/backends/git-gateway/implementation.js]: indentation

File: src/backends/git-gateway/implementation.js [hunk 4]
```diff
@@ -98,6 +109,14 @@ export default class GitGateway {
...
+          ...apiConfig,
+          requestFunction: this.requestFunction,
+          hasWriteAccess: async () => true,
+        });
+        console.log(this.api);
+	this.backend = new BitBucketBackend(this.config, { proxied: true, API: this.api });
```
Comment (erquhart) [inline src/backends/git-gateway/implementation.js]: indentation lol. obviously due to your stubborn use of emacs. also we need prettier.

File: src/lib/backendHelper.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    return formatter(res);
+  } catch (err) {
+    throw new Error(`Response cannot be parsed into the expected format (${ format }): ${ err.message }`);
+  }
+};
+
+const responseFormatters = fromJS({
+  json: async res => {
+    const contentType = res.headers.get("Content-Type");
+    if (!contentType.startsWith("application/json") && contentType.startsWith("text/json")) {
```
Comment (erquhart) [inline src/lib/backendHelper.js]: This can be switched to just check for `startsWith('text/json')`.

File: src/backends/bitbucket/API.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    const branch = options.branch || this.branch;
+    const body = new FormData();
+    body.append('files', path);
+    if (message && message !== "") {
+      body.append("message", message);
+    }
+    return flow([
+      unsentRequest.withMethod("POST"),
+      unsentRequest.withBody(body),
+      this.request,
```
Comment (erquhart) [inline src/backends/bitbucket/API.js]: _Love_ this pattern for building requests, awesome stuff.

File: src/lib/backendHelper.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  text: async res => res.text(),
+  blob: async res => res.blob(),
+}).mapEntries(
+  ([format, formatter]) => [format, catchFormatErrors(format, formatter)]
+);
+
+export const parseResponse = async (res, { expectingOk = true, format = "text" } = {}) => {
+  if (expectingOk && !res.ok) {
+    throw new Error(`Expected an ok response, but received an error status: ${ res.status }.`);
+  }
```
Comment (erquhart) [inline src/lib/backendHelper.js]: Just noting that this triggers UI errors in situations that don't currently show errors in other backends. For example, a collection in the configuration whose folder does not exist in the repo shows this error, while failing silently in other backends.

File: src/lib/backendHelper.js (hunk 5 above)
Comment (Benaiah) [reply src/lib/backendHelper.js]: It should actually be `!contentType.startsWith("application/json") && !contentType.startsWith("text/json")`, but the second `!` was missing. Parsing a `Content-Type` of `"text/json"` as JSON shouldn't produce an error.

File: src/backends/git-gateway/implementation.js (hunk 4 above)
Comment (Benaiah) [reply src/backends/git-gateway/implementation.js]: :+1: to getting prettier in

**Pull request merged**